File: 125-valid-palindrome/valid-palindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        s = s.toLowerCase();
        int i = 0, j = s.length() - 1;
        while (i <= j) {
            while (i <= j && !isAlphanumeric(s, i)) i++;
            while (i <= j && !isAlphanumeric(s, j)) j--;
            if (i <= j) {
                if(s.charAt(i) == s.charAt(j)){
                    i++;
                    j--;
                }
                else return false;
            } 
        }
        return true;
    }

    public boolean isAlphanumeric(String s, int p) {
        return (s.charAt(p) >= 'a' && s.charAt(p) <= 'z') || (s.charAt(p) >= '0' && s.charAt(p) <= '9');
    }
}

File: 125-valid-palindrome/valid-palindrome.py
class Solution(object):
    def isPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """
        lo = 0 
        hi = len(s) - 1
        s = s.lower()
        while lo < hi:
            while lo < hi and not self.isAlphanumeric(s[lo]):
                lo += 1
            while lo < hi and not self.isAlphanumeric(s[hi]):
                hi -= 1
            if lo >= hi:
                return True
            elif s[lo] == s[hi]:
                lo += 1
                hi -= 1
            else:
                return False
        return True
    def isAlphanumeric(self, c):
        return '0'<=c<='9' or 'a'<=c<='z'
